hydra-node/src/Hydra/Network/Reliability.hs

```diff
@@ -0,0 +1,294 @@
+                broadcast $ ReliableMsg ackCounter' msg
+              Ping{} -> do
+                acks <- readTVarIO ackCounter
+                traceWith tracer (BroadcastPing ourIndex acks)
+                broadcast $ ReliableMsg acks msg
+        }
+
+  reliableCallback ackCounter sentMessages seenMessages resend (Authenticated (ReliableMsg acks msg) party) = do
+    if length acks /= length allParties
+      then ignoreMalformedMessages
+      else do
+        partyIndex <- findPartyIndex party
+        (shouldCallback, messageAckForParty, knownAckForParty, knownAcks) <- atomically $ do
+          let messageAckForParty = acks ! partyIndex
+          knownAcks <- readTVar ackCounter
+          let knownAckForParty = knownAcks ! partyIndex
+
+          -- handle message from party iff it's next in line OR if it's a Ping
+          --
+          -- The stream of messages from a peer is expected to look like:
+          --
+          -- @@
+          -- Msg [.., 1, ..] (Data nid m1)
+          -- Msg [.., 2, ..] (Data nid m1)
+          -- Msg [.., 3, ..] (Data nid m1)
+          -- Msg [.., 3, ..] (Ping nid)
+          -- Msg [.., 3, ..] (Ping nid)
+          -- Msg [.., 3, ..] (Ping nid)
+          -- @@
+          --
+          -- Pings are observed only for the information it provides about the
+          -- peer's view of our index
+          if messageAckForParty == knownAckForParty + 1
+            then do
+              let newAcks = constructAcks knownAcks partyIndex
+              writeTVar ackCounter newAcks
+              return (True, messageAckForParty, knownAckForParty, newAcks)
+            else return (isPing msg, messageAckForParty, knownAckForParty, knownAcks)
+
+        if shouldCallback
+          then do
+            callback (Authenticated msg party)
+            traceWith tracer (Received acks knownAcks partyIndex)
+          else traceWith tracer (Ignored acks knownAcks partyIndex)
+
+        -- resend messages if party did not acknowledge our latest idx
+        resendMessages resend partyIndex sentMessages knownAcks acks messageAckForParty knownAckForParty
+
+        -- Update last message index sent by us and seen by some party
+        updateSeenMessages seenMessages acks party
+        -- Take the lowest number from seen messages by everyone and remove it from
+        -- our sent messages.
+        deleteSeenMessages sentMessages seenMessages
+
+  ignoreMalformedMessages = pure ()
+
+  constructAcks acks wantedIndex =
+    zipWith (\ack i -> if i == wantedIndex then ack + 1 else ack) acks partyIndexes
+
+  partyIndexes = generate (length allParties) id
+
+  resendMessages resend partyIndex sentMessages knownAcks messageAcks messageAckForParty knownAckForParty = do
+    myIndex <- findPartyIndex me
+    let messageAckForUs = messageAcks ! myIndex
+    let knownAckForUs = knownAcks ! myIndex
+    when (messageAckForUs < knownAckForUs && messageAckForParty <= knownAckForParty) $ do
```

Comment:
This `messageAckForParty <= knownAckForParty` decision can be taken outside of this and is even documented on line 221. I would suggest to move it there and have it read:

```
        when (messageAckForParty <= knownAckForParty) $
          resendMessages resend partyIndex sentMessages knownAcks acks
```